02-household-queries/chainlit-household-bot.py

```diff
@@ -182,3 +236,57 @@ def call_llm(message: cl.Message):
     client = cl.user_session.get("client")
     response = client.invoke(message.content)
     return response
+
+@cl.action_callback("uploadQueryAct")
+async def on_click_upload_file_query(action: cl.Action):
+    embeddings = cl.user_session.get("embedding")
+    llm = cl.user_session.get("client")
+    files = None
+    # Wait for the user to upload a file
+    while files == None:
+        files = await cl.AskFileMessage(
+            content="Please upload a pdf or json file to begin!",
+            accept=["text/plain", "application/pdf", "application/json"],
+            max_size_mb=20,
+            timeout=180,
+        ).send()
+    file = files[0]
+    
+    # initialize db
+    vectordb=Chroma(embedding_function=embeddings, collection_name="resources", persist_directory="./chroma_db")
+    cl.user_session.set("vectordb", vectordb)
+
+    if(file.type == "application/pdf"):
+        add_pdf_to_vector_db(vectordb=vectordb, file_path=file.path)
+    elif(file.type == "application/json"):
+        add_json_html_data_to_vector_db(vectordb=vectordb, file_path=file.path, content_key="content", index_key="preferredPhrase")
+    msg = cl.Message(content=f"Processing `{file.name}`...", disable_feedback=True)
+    await msg.send()
+    
+    retriever = vectordb.as_retriever(search_kwargs={"k": 1})
+    message_history = ChatMessageHistory()
+
+    memory = ConversationBufferMemory(
+        memory_key="chat_history",
+        output_key="answer",
+        chat_memory=message_history,
+        return_messages=True,
+    )
+
+    # Create a chain that uses the Chroma vector store
+    chain = ConversationalRetrievalChain.from_llm(
+        llm,
+        chain_type="stuff",
+        retriever=retriever,
+        memory=memory,
+        return_source_documents=True,
+    )
+    
+    
+    # Let the user know that the system is ready
+    msg.content = f"Processing `{file.name}` done. You can now ask questions!"
+    await msg.update()
+
+    cl.user_session.set("chain", chain)
```

Comment:
Refactor into separate function, and call the function within the `message_submitted` function. Otherwise, the wrong llm may be used if the llm is chosen after this action function is run.